stl/inc/xstring

```diff
@@ -2463,23 +2512,94 @@ private:
             int>;
 #endif // _HAS_CXX17
 
-public:
-    _CONSTEXPR20 basic_string(const basic_string& _Right)
-        : _Mypair(_One_then_variadic_args_t{}, _Alty_traits::select_on_container_copy_construction(_Right._Getal())) {
-        auto&& _Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
-        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _Mypair._Myval2);
-        _Construct_lv_contents(_Right);
-        _Proxy._Release();
+#ifdef _INSERT_STRING_ANNOTATION
+    _CONSTEXPR20 void _Create_annotation() const noexcept {
+        // Annotates the valid range with shadow memory
+        auto& _My_data = _Mypair._Myval2;
+        _Apply_annotation(_My_data._Myptr(), _My_data._Myres, _My_data._Myres, _My_data._Mysize);
     }
 
-    _CONSTEXPR20 basic_string(const basic_string& _Right, const _Alloc& _Al)
-        : _Mypair(_One_then_variadic_args_t{}, _Al) {
-        auto&& _Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
-        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _Mypair._Myval2);
-        _Construct_lv_contents(_Right);
-        _Proxy._Release();
+    _CONSTEXPR20 void _Remove_annotation() const noexcept {
+        // Removes annotation of the range with shadow memory
+        auto& _My_data = _Mypair._Myval2;
+        _Apply_annotation(_My_data._Myptr(), _My_data._Myres, _My_data._Mysize, _My_data._Myres);
+    }
+
+    _CONSTEXPR20 void _Modify_annotation(const difference_type _Count) const noexcept {
+        // Extends/shrinks the annotated range by _Count
+        auto& _My_data = _Mypair._Myval2;
+        _Apply_annotation(
+            _My_data._Myptr(), _My_data._Myres, _My_data._Mysize, static_cast<size_type>(_My_data._Mysize + _Count));
+    }
+
+    _NODISCARD static const void* _Get_aligned_first(const void* _First, const size_type _Capacity) noexcept {
+        const char* _CFirst = reinterpret_cast<const char*>(_First);
+
+        if (_Capacity >= _Asan_granularity) { // We are guaranteed to have sufficient space to find an aligned address
+            return reinterpret_cast<void*>(
+                (reinterpret_cast<uintptr_t>(_CFirst) + (_Asan_granularity - 1)) & ~(_Asan_granularity - 1));
+        }
+
+        uintptr_t _Alignment_offset = reinterpret_cast<uintptr_t>(_CFirst) & (_Asan_granularity - 1);
+        if (_Alignment_offset != 0) {
+            _Alignment_offset = _Asan_granularity - _Alignment_offset;
+        }
+
+        if (_Capacity > _Alignment_offset) {
+            return _CFirst + _Alignment_offset;
+        }
+
+        return nullptr;
+    }
+
+    static _CONSTEXPR20 void _Apply_annotation(const value_type* _Ptr, const size_type _Capacity,
+        const size_type _Old_size, const size_type _New_size) noexcept {
+#if _HAS_CXX20
+        if (_STD is_constant_evaluated()) {
+            return;
+        }
+#endif // _HAS_CXX20
+
+        if (!_Asan_string_should_annotate) {
+            return;
+        }
+
+        if (_Capacity == _BUF_SIZE - 1) { // TRANSITION, DevCom-1527920

```

Comment:
I'm still of the opinion that we should enable annotations of the SSO buffer and just live with the fact that any exception getting thrown will just clobber it.

Reply:
Yeah, I am leaning towards it. The beauty is that we only need to remove that if and everything should already work

Reply:
Enabled the optimizations